File: 2015/11/solution.js
const validPass = (password) => {
  return !!password.match(/(abc)|(bcd)|(cde)|(def)|(efg)|(fgh)|(pqr)|(qrs)|(rst)|(stu)|(tuv)|(uvw)|(vwx)|(wxy)|(xyz)/)
      && !password.match(/[iol]/)
      && !!password.match(/(.)\1.*(.)\2/)
}

const nextChar = (str, pos) => String.fromCharCode(str.charCodeAt(pos) + 1)

const replaceChar = (str, pos, char) => str.substr(0, pos) + char + str.substr(pos + 1)

const incrementPass = (password) => {
  let next = password
  let pos = password.length - 1
  while (true) {
    if (next.charAt(pos) === 'z') {
      next = replaceChar(next, pos, 'a')
      pos -= 1
    } else {
      return replaceChar(next, pos, nextChar(next, pos))
    }
  }
}

const solve = (input) => {
  let password = input
  while (true) {
    password = incrementPass(password)
    if (validPass(password)) {
      return password
    }
  }
}

const pass1 = solve(require('fs').readFileSync('input.txt', 'utf8'))
const pass2 = solve(pass1)

console.log('Part 1', pass1)
console.log('Part 2', pass2)

File: 2015/11/solution.rb
def valid_pass(password)
  !!password.match(/(abc)|(bcd)|(cde)|(def)|(efg)|(fgh)|(pqr)|(qrs)|(rst)|(stu)|(tuv)|(uvw)|(vwx)|(wxy)|(xyz)/) &&
      !password.match(/[iol]/) &&
      !!password.match(/(.)\1.*(.)\2/)
end

def increment_pass(password)
  pos = password.size - 1
  while true do
    if password[pos] == 'z'
      password[pos] = 'a'
      pos -= 1
    else
      password[pos] = (password[pos].ord + 1).chr
      return
    end
  end
end

def solve(input)
  password = input
  while true do
    increment_pass(password)
    return password if valid_pass(password)
  end
end

puts 'Part 1', pass = solve(File.read('input.txt'))
puts 'Part 2', solve(pass)
